### PR TITLE
Feat/2854 sign verify

### LINKED ISSUE #2854: Sign & verify 🔏
**Design:** [Zeplin](https://zpl.io/aMrekQz)

Sign and verify messages and addresses.

### PR DESCRIPTION
**Pull request by marekrjpolak**
Adds Sign&Verify functionality (#2854), UI mostly based on [Zeplin](https://app.zeplin.io/project/5ee87c00f6af719a645d14c3/screen/5faa87d7aab597690780cb19). Demonstration [here](https://suite.corp.sldev.cz/suite-web/feat/2854-sign-verify/web/).

- d3ec738 - Repeated header code (title on the left, close button and some actions on the right) separated from wallet pages to `WalletLayoutHeader` component and used in Sign&Verify page.
- 9d37edf - Tab navigation from `CoinMarketLayout` isolated to `WalletLayoutNavigation` component so it could be used in Sign&Verify page.
- ddbd079 - `signVerifyReducer` removed and `signVerifyActions` (re)implemented to support signing, verifying and address checking for all suported coins.
- c7c536e - Added `SignAddressInput` for selecting signing address. The logic for obtaining addresses by selected coin, grouping them (fresh/used) and selecting them is implemented in `useSignAddressOptions` hook. `VerifyAddressButton` enables to reveal/check the address, because they are partially hidden by default.
- d1a54aa - Implemented Sign&Verify page with form and all requested inputs, with validation as needed.
- 2f96f5e - Sign&Verify option shown in menu for all supported coins.
- 0079afc - Added copy to clipboard functionality in `useCopySignedMessage` hook.
- 5819832 - `react-hook-form` logic separated from UI and wrapped by `useSignVerifyForm` hook.

### CODE REVIEW

File: packages/suite/src/views/wallet/sign-verify/index.tsx [hunk 1]
```diff
@@ -1,207 +1,170 @@
...
-                                        signVerify.verifyAddress,
-                                        signVerify.verifyMessage,
-                                        signVerify.verifySignature,
-                                    );
-                                }
-                            }}
-                        >
-                            <Translation id="TR_VERIFY" />
+                    <Row>
+                        <StyledButton type="submit">
```
Comment (mroz22) [inline packages/suite/src/views/wallet/sign-verify/index.tsx]: not a big deal but I believe that this button should be disabled if there is still some kind of communication with device in progress. Possible edge case: 
1. enable all coins on all all seed (or any other larger account)
2. reload page 
3. quickly try to sign some message
4. get error "device call in progress"

I think there is `useDevice` hook which may help with this.

**Review by mroz22 — CHANGES_REQUESTED** [review]

File: packages/suite/src/views/wallet/sign-verify/index.tsx (hunk 1 above)
Comment (mroz22) [reply packages/suite/src/views/wallet/sign-verify/index.tsx]: live example in #4300

Comment (marekrjpolak) [reply packages/suite/src/views/wallet/sign-verify/index.tsx]: Agreed. Buton disabling implemented, e2e tests merged and extended. Everything seems to be working.

File: patches/trezor-connect+8.1.30-beta.7.patch [hunk 2]
```diff
@@ -0,0 +1,20 @@
+diff --git a/node_modules/trezor-connect/lib/typescript/networks/bitcoin.d.ts b/node_modules/trezor-connect/lib/typescript/networks/bitcoin.d.ts
+index 0c4ad87..66b0ffb 100644
+--- a/node_modules/trezor-connect/lib/typescript/networks/bitcoin.d.ts
++++ b/node_modules/trezor-connect/lib/typescript/networks/bitcoin.d.ts
```
Comment (szymonlesisz) [inline patches/trezor-connect+8.1.30-beta.7.patch]: rebase over develop should help, this patch is no longer necessary

**Comment by szymonlesisz** [comment]
use `network.features` ('sign-verify') instead of `networks.hasSignVerify`

File: packages/suite/src/components/wallet/WalletLayout/components/WalletLayoutNavigation.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    soon?: boolean;
+};
+
+export const WalletLayoutNavLink = ({ active, title, onClick, soon }: WalletLayoutNavLinkProps) => (
+    <NavLink active={active} onClick={() => !soon && onClick()}>
+        <NavLinkText>
+            {soon && (
+                <Soon>
+                    <Translation id="TR_NAV_SOON" />
+                </Soon>
```
Comment (tomasklim) [inline packages/suite/src/components/wallet/WalletLayout/components/WalletLayoutNavigation.tsx]: I would remove all `soon` code as it is not used anymore.

File: packages/suite/src/hooks/wallet/sign-verify/useSignAddressOptions.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import type { State as RevealedAddresses } from '@wallet-reducers/receiveReducer';
+
+export type AddressItem = {
+    label: string;
+    value: string;
+};
+
+export const useSignAddressOptions = (
+    account: Account | undefined,
+    revealedAddresses: RevealedAddresses,
```
Comment (tomasklim) [inline packages/suite/src/hooks/wallet/sign-verify/useSignAddressOptions.ts:14]: ```revealedAddresses``` is required only if `account` is defined

**Comment by matejkriz** [comment]
The description of the commits here is really helpful and make code review easier, what about to put them next time directly in commit messages? It could be useful for future reference (easy to access from git blame).

 But of course do it only when it is not too much extra work, I understand that you probably wrote this descriptions retrospectively...

**Comment by matejkriz** [comment]
Could we close this one now? [sign & verify in bitcoin and ethereum networks](https://github.com/trezor/trezor-suite/pull/2020) https://github.com/trezor/trezor-suite/pull/2020

File: packages/suite/src/hooks/wallet/sign-verify/useSignVerifyForm.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
+import { useEffect } from 'react';
+import { useForm, useController } from 'react-hook-form';
+import { useTranslation } from '@suite-hooks';
+import { isHex } from '@wallet-utils/ethUtils';
+import { isAddressValid } from '@wallet-utils/validation';
+import type { Account } from '@wallet-types';
+
+export const MAX_LENGTH_MESSAGE = 255;
+export const MAX_LENGTH_SIGNATURE = 255;
```
Comment (tomasklim) [inline packages/suite/src/hooks/wallet/sign-verify/useSignVerifyForm.ts:9]: We can display how many chars is left as in guide > report bug text area

File: packages/suite/src/views/wallet/sign-verify/index.tsx [hunk 6]
```diff
@@ -1,207 +1,180 @@
...
+                                type="text"
+                                state={errorState(formErrors.address)}
+                                bottomText={formErrors.address}
+                                data-test="@sign-verify/select-address"
+                                {...addressField}
+                            />
+                        )}
                     </Row>
                     <Row>
                         <Textarea
```
Comment (matejkriz) [inline packages/suite/src/views/wallet/sign-verify/index.tsx:160]: What about adding the placeholder `Will be generated after Signing` as it is in the design?
image

File: packages/suite/src/hooks/wallet/sign-verify/useSignAddressOptions.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    const signAddresses = useMemo(() => {
+        switch (account?.networkType) {
+            case 'bitcoin':
+                return {
+                    ...reduceAddresses(
+                        revealedAddresses.length
+                            ? revealedAddresses
+                            : (account.addresses?.unused || []).slice(0, 1),
+                        'TR_ADDRESSES_FRESH',
```
Comment (matejkriz) [inline packages/suite/src/hooks/wallet/sign-verify/useSignAddressOptions.ts:40]: nit: could be singular `Fresh address`

**Review by tomasklim — CHANGES_REQUESTED** [review]
1. If you switch to hex format > Clear all > Error is still there.
Screenshot 2021-09-30 at 11 41 51

2. If you select address and you click on Address again, the cursor is too far in right
Screenshot 2021-09-30 at 11 42 15

3. You are in Sign message and you click on Hex format switch and you navigate to Verify message, an error is still there

4. If you scroll a little bit, a section is missing in the sticky header.
Screenshot 2021-09-30 at 11 51 29

5. I would move the Copy button as you cannot see it when you scroll and I also did not notice that it appeared there.
Screenshot 2021-09-30 at 11 53 05

6. You will probably rework it but space is missing there
Screenshot 2021-09-30 at 11 54 18

7. I am not sure if notification is enough I would expect e.g. green Signature textbox or some green/red text in the form
Screenshot 2021-09-30 at 11 55 05

8. Notification is missing when you sign a message

File: packages/suite/src/hooks/wallet/sign-verify/useSignAddressOptions.ts (hunk 7 above)
Comment (marekrjpolak) [reply packages/suite/src/hooks/wallet/sign-verify/useSignAddressOptions.ts:40]: If you reveal more addresses in Receive form before, you can then sign with any of them:
image

File: patches/trezor-connect+8.1.30-beta.7.patch (hunk 2 above)
Comment (marekrjpolak) [reply patches/trezor-connect+8.1.30-beta.7.patch]: https://github.com/trezor/trezor-suite/pull/4297/commits/9c16607512af09a5f8125a6fd7ccb7fb72e15127

**Comment by marekrjpolak** [comment]
> use `network.features` ('sign-verify') instead of `networks.hasSignVerify`

https://github.com/trezor/trezor-suite/pull/4297/commits/2235ce426ba5aee9da0c6204efdae3a524489229

File: packages/suite/src/views/wallet/sign-verify/index.tsx (hunk 6 above)
Comment (marekrjpolak) [reply packages/suite/src/views/wallet/sign-verify/index.tsx:160]: https://github.com/trezor/trezor-suite/pull/4297/commits/43a8c4291aa250e4aa8a537b7de7e38581948364

File: packages/suite/src/hooks/wallet/sign-verify/useSignVerifyForm.ts (hunk 5 above)
Comment (marekrjpolak) [reply packages/suite/src/hooks/wallet/sign-verify/useSignVerifyForm.ts:9]: https://github.com/trezor/trezor-suite/commit/4f59290265ca895e38f592846936fc9157c3c6df

File: packages/suite/src/components/wallet/WalletLayout/components/WalletLayoutNavigation.tsx (hunk 3 above)
Comment (marekrjpolak) [reply packages/suite/src/components/wallet/WalletLayout/components/WalletLayoutNavigation.tsx]: https://github.com/trezor/trezor-suite/commit/6dd0f6393aeea25d520a92a92a46a76648feb823

**Comment by marekrjpolak** [comment]
> 1. If you switch to hex format > Clear all > Error is still there.
> 3. You are in Sign message and you click on Hex format switch and you navigate to Verify message, an error is still there

https://github.com/trezor/trezor-suite/commit/5f4bcbbac4f0134a9720ee1adefe0e1a2a87177a
> 2. If you select address and you click on Address again, the cursor is too far in right

https://github.com/trezor/trezor-suite/commit/47075d3196e43caecf7a2265f3012a86d3a47604
> 4. If you scroll a little bit, a section is missing in the sticky header.

https://github.com/trezor/trezor-suite/commit/0666f35a1d3dbdc58aea68270cad3593bcbea255
> 6. You will probably rework it but space is missing there

https://github.com/trezor/trezor-suite/commit/8bab9623b10a1d430e577b37dc20375a432a4e1a
> 8. Notification is missing when you sign a message

https://github.com/trezor/trezor-suite/commit/3eb9add79f9ca8dfcb32458804c59e8ff966f6d9
> 5. I would move the Copy button as you cannot see it when you scroll and I also did not notice that it appeared there.
> 7. I am not sure if notification is enough I would expect e.g. green Signature textbox or some green/red text in the form

I'll ask product team what to do with it.

**Review by tomasklim — DISMISSED** [review]
- You can activate "Hex format" switch by clicking on "Message" label
- When you activate "Hex format" when text area is blank you got an error and this error does not disappear when you type correct hex. It disappears only when you unfocus of the text area
https://user-images.githubusercontent.com/33235762/135815185-c33ae91f-937d-4c55-836e-bd79aa05957f.mov

**Comment by marekrjpolak** [comment]
> * You can activate "Hex format" switch by clicking on "Message" label

https://github.com/trezor/trezor-suite/pull/4297/commits/95df175bd5d2dd32716833226f79888a39c03805
> * When you activate "Hex format" when text area is blank you got an error and this error does not disappear when you type correct hex. It disappears only when you unfocus of the text area

https://github.com/trezor/trezor-suite/pull/4297/commits/fe63a5e98cb340a0a8c87710056b69010a8aac00

**Review by mroz22 — APPROVED** [review]

**Review by mroz22 — APPROVED** [review]

**Pull request merged**